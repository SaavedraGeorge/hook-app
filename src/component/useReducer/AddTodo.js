import React from 'react';
import { useForm } from '../../hooks/useForm';

export const AddTodo = ({handleAddTodo}) => {
    const [{description}, handleInputChange, reset] = useForm({ 
        description: ''
     });
    
    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1){
            return
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false 
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar To Do</h4>
            <hr />
            <form onSubmit={ handleSubmit }>
                <input 
                    type="text" 
                    name="description" 
                    placeholder="add To Do" 
                    autoComplete='off'
                    className='form-control'
                    value={ description }
                    onChange={ handleInputChange }
                />
                <button 
                    className='btn btn-outline-primary col-12 mt-1'
                    type='submit'
                >Agregar</button>
            </form>
        </>
    );
};
