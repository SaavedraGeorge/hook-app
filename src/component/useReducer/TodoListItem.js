import React from 'react';

export const TodoListItem = ({todo, i, handleDelete, toggle }) => {
  
  return (
    <li
    className='list-group-item'
    key={todo.id}
    >
        <p 
            onClick={ () => toggle(todo.id)}
            className={`${ todo.done ? 'complete': ''}`}
        >
            {i+1}. { todo.desc }
        </p>
        <button 
            className='btn btn-danger'
            onClick={ () => handleDelete(todo.id) }
        >Eliminar</button>
    </li>
  );
};
