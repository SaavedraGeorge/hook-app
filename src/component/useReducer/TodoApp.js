import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';


import './style.css'
import { TodoList } from './TodoList';
import { AddTodo } from './AddTodo';

const init = () =>{
    
    // llama lo que hay en el loca storrage y si es vacio devuelve []
    return JSON.parse(localStorage.getItem('todos')) || [];
    
} 

export const TodoApp = () => {

    const [todos, dispatch] = useReducer( todoReducer, [], init );

    
    
    
     useEffect(() => {
       localStorage.setItem('todos', JSON.stringify(todos));
     
     }, [todos]);
     

    const handleAddTodo = ( newTodo ) => {
        
        dispatch({
            type: 'add',
            payload: newTodo
        });

    }

    const handleDelete = (idItem) => {
        
        const action ={
            type: 'delete',
            payload: idItem
        }

        dispatch(action);

    }

    const toggle = (idItem) => {
        dispatch( {
            type: 'toggle',
            payload: idItem 
        } )
    }

    return (
        <div>
            <h1>To Do App <small>{todos.length}</small></h1>
            <hr />

  

            <div className='row'>
                <div className='col-7'>
                    <TodoList 
                        todos = {todos}
                        handleDelete = {handleDelete}
                        toggle = {toggle}
                    /> 
                </div>
                <div className='col-5'>
                    <AddTodo handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}
