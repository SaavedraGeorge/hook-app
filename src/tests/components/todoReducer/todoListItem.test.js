import React from 'react';
import { shallow } from "enzyme";
import { TodoListItem } from '../../../component/useReducer/TodoListItem';
import { demoTodo } from '../../fixtures/demoTodo';



describe('pruebas a <TodoReducer />', () => {
    const handleDelete = jest.fn();
    const toggle = jest.fn(); 
    
    const wrapper = shallow(
        <TodoListItem
            todo = {demoTodo[0]} 
            i = {0}
            handleDelete = {handleDelete}
            toggle = {toggle} 
        />
    );

    test('should snapshot ', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('handDelete tests', () => {
      wrapper.find('button').simulate('click');
      //entrega el id del elemento a borrar
      expect( handleDelete ).toHaveBeenCalledWith(demoTodo[0].id);
    });
    test('handDelete toggle', () => {
        wrapper.find('p').simulate('click');
        //entrega el id del elemento a borrar
        expect( toggle ).toHaveBeenCalledWith(demoTodo[0].id);
      });
      test('should show the text', () => {
        const p = wrapper.find('p');
        expect(p.text().trim()).toBe(`1. ${demoTodo[0].desc}`)
      });
      test('should show complete class', () => {
        const todo = demoTodo[0];
        //le asignamos done true al elemento 0 del arreclo
        todo.done = true;

        //llamamos al wrapper con el nuevo todo
        const wrapper = shallow(<TodoListItem
            todo = {todo} />)
        
        //buscamos el parrafo
        const p = wrapper.find('p');

        // vemos si el parrafo tiene la clase 'complete'
        expect(p.hasClass('complete')).toBe(true);
      });
      
      
    


});
