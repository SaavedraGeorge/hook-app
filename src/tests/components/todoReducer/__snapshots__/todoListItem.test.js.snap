// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas a <TodoReducer /> should snapshot  1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className=""
    onClick={[Function]}
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Eliminar
  </button>
</li>
`;
